### PR TITLE
Deprecate `Matrix::SaveType::Full`

### PR DESCRIPTION
**Pull request by JonathonMisiewicz**
## Description
While working on some more DF technology, I discovered a bug when trying to load using `Matrix::SaveType::Full`. The code assumed the matrix was square. One of the things this PR does is have the code work when that assumption fails.

But the bigger problem is that `Matrix::SaveType::Full` is Psi3 era tech that shouldn't exist anymore. That SaveType reads/writes a matrix _where all elements of the wrong symmetry are replaced with zeros_. The better solution is to use `Matrix::SaveType::SubBlocks`, where those zero's are neither written nor read. Accordingly, I'm deprecating the Full save in favor of SubBlocks save.

## Todos

- [x] `Matrix::SaveType::Full` is deprecated
- [x] Fixed a correctness bug with loading from `Matrix::SaveType::Full`

## Checklist
- [x] quicktests, standard stuite, and dfmp2 all pass
- [ ] I didn't bother adding a new test that I actually fixed the bug, since I accessed the buggy code through a method that's going to be deprecated anyways. Any objections?

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Review by loriab — APPROVED** [review]

**Comment by loriab** [comment]
It looks like gcc5.5 has a grievance with the deprecated syntax https://dev.azure.com/psi4/psi4/_build/results?buildId=2687&view=logs&j=258bfeeb-871e-502d-292b-61cd4e6d17f6&t=7c2e2bd2-871d-5948-1d43-9089e6bce09a&l=253 . Looks like it came about in gcc6 https://stackoverflow.com/a/33096368 . I can confirm gcc 7.3 is ok with it.

What do we want to do? Give up the deprecated enum value or bump the gcc requirement?

**Comment by JonathonMisiewicz** [comment]
I'm inclined to force an update, given that GCC5 hasn't been supported for years, but as I'm not familiar with cluster management, my opinion isn't very strong.

**Comment by loriab** [comment]
fwiw, I don't anticipate GT having problems if GCC min boosted. alternately, can do alternate defs of the enum with preprocessor variables.

**Comment by PeterKraus** [comment]
I think Debian Jessie, which is still in Extended LTS period, ships with GCC 4.9. Stretch (which is in LTS) ships with 6.3. Similarly, it looks like everything older than (and including) RHEL7 came with GCC 4.8, but for RHEL6 and RHEL7 their respective DTS packages come with GCC 6.2 or newer.

I vote to deprecate GCC 5 on principle...

**Review by jturney — APPROVED** [review]

File: psi4/src/psi4/libmints/matrix.cc [hunk 1]
```diff
@@ -3037,7 +3036,7 @@ void Matrix::load(psi::PSIO *const psio, size_t fileno, SaveType st) {
             psio->read_entry(fileno, name_.c_str(), (char *)fullblock[0], sizeof(double) * sizer * sizec);
 
         set(fullblock);
-        linalg::detail::free(fullblock);
```
Comment (robertodr) [inline psi4/src/psi4/libmints/matrix.cc:3040]: Why the change?

**Review by robertodr — APPROVED** [review]

File: psi4/src/psi4/libmints/matrix.cc (hunk 1 above)
Comment (JonathonMisiewicz) [reply psi4/src/psi4/libmints/matrix.cc:3040]: LGTM was reporting a mismatch error. I didn't dig into the innards of `linalg::detail::free`, but `freeblock` seems to be the standard way to clear memory allocated with `to_block_matrix`.

**Comment by loriab** [comment]
if you rebase now, Azure should be healed

**Comment by JonathonMisiewicz** [comment]
LGTM seems to be having problems. Not sure if that means we should delay merging this in until we understand why or not.

I need this PR in before I can continue on a bugfix PR, so I'd prefer to have this in ASAP.

**Comment by loriab** [comment]
I'm expecting C++ lgtm to fail for the foreseeable future b/c the additional analysis of libxc is exceeding the 4h time limit. Not sure why it's running out of disk space <3h, but probably not the fault of this PR. will merge.
